Support for non-immediate proof maintenance

PetarMax opened the pull request:
When doing large real-world proofs with a high degree of parallelism, the bandwidth is overtaken by the writing of the proof to disk on every proof iteration, slowing the process considerably.

This PR introduces the ability to write to disk and perform the callback on every `maintenance_iterations` of the parallel proof, set by default to 32.

EDIT: Actually, perhaps it's a better idea to write the proof every `max_workers` iterations.

Commit:
Support for non-immediate proof maintenance

Commit:
maintenance every max_workers iterations

rv-jenkins committed:
Update dependency: deps/llvm-backend_release (#4559)

Co-authored-by: devops <devops@runtimeverification.com>

nwatson22 committed:
Display number of terminal nodes in `one_line_summary` for `APRProof` (#4566)

Adds the number of terminal nodes to `one_line_summary` for `APRProof`.
Also changes the display to use `|` as dividers instead of `/`.

rv-jenkins committed:
Update dependency: deps/haskell-backend_release (#4558)

Co-authored-by: devops <devops@runtimeverification.com>

rv-jenkins committed:
Update dependency: deps/haskell-backend_release (#4571)

Co-authored-by: devops <devops@runtimeverification.com>

nwatson22 reviewed (approved):
Looks good. Can you make it toggleable with a flag so we can choose to do the disk writing/status bar updating every step?

geo2a committed:
Remove the deprecated `"log-timing"` RPC log option (#4573)

This option is unused and all its use cases are subsumed by the `-l
Timing --log-format json` non-RPC log option.

For reference, here's the Haskell Backend PR that stops emitting the
timing RPC log entries:
https://github.com/runtimeverification/haskell-backend/pull/4015

PetarMax commented:
I think I prefer to have a separate parameter so that we can decouple it from the number of workers, as it is now. Is there anything else to do in `pyk`?

PetarMax committed:
Merge branch 'develop' into petar/non-immediate-proof-maintenance

Commit:
parameter description

nwatson22 commented:
> I think I prefer to have a separate parameter so that we can decouple it from the number of workers, as it is now. Is there anything else to do in `pyk`?

You could possibly add the same parameter/logic to `Prover.advance_proof()` and feed the option through to be able to be used with `pyk prove`, but that could also be done later in a separate PR since this is needed for kontrol now.

nwatson22 commented on pyk/src/tests/unit/test-data/pyk_toml_test.toml:
```diff
@@ -1,6 +1,6 @@
 [coverage]
 output = "default-file"
-definition = "/tmp"
+definition = "/var/folders/ks/1p8zyp5j7xz_v1krhl4l17tm0000gn/T"
```
What is this change for?

PetarMax replied:
I have not changed that file manually...

Commit:
reverting accidental change

Commit:
adding option to Prover.advance_proof

PetarMax commented:
> > I think I prefer to have a separate parameter so that we can decouple it from the number of workers, as it is now. Is there anything else to do in `pyk`?
> 
> You could possibly add the same parameter/logic to `Prover.advance_proof()` and feed the option through to be able to be used with `pyk prove`, but that could also be done later in a separate PR since this is needed for kontrol now.

Yes, happy to.

PetarMax commented:
Just to say - I am not sure how to test this, because the final outcome is the same, it's just that the proof should be faster. Happy to go along with any ideas in a follow-up PR.
